### PR TITLE
Refactored tracing strategy, fixes #29

### LINKED ISSUE #29: Possible memory leak - datasource proxy
Hello, today I have come across possible memory leak in datasource proxy inside `TracingListenerStrategy`

After 250k requests to server, the strategy instance acumulates nearly 900MB of memory and never releases it.

image

I have latest version of datasource proxy spring boot starter. 

```
        <dependency>
            <groupId>com.github.gavlyukovskiy</groupId>
            <artifactId>datasource-proxy-spring-boot-starter</artifactId>
            <version>1.5.3</version>
        </dependency>
```

configuration properties

```
decorator:
  datasource:
    datasource-proxy:
      query:
        enable-logging: false
      slow-query:
        threshold: 1
      count-query: true
```
When using p6spy, the problem does not seem to be present.

### PR DESCRIPTION
**Pull request by gavlyukovskiy**
Refactored tracing strategy to have connection as key for tracking all related statements and result sets.

### CODE REVIEW

File: datasource-decorator-spring-boot-autoconfigure/src/main/java/com/github/gavlyukovskiy/cloud/sleuth/TracingListenerStrategy.java [hunk 1]
```diff
@@ -30,136 +21,203 @@ void beforeGetConnection(CON connectionKey, String dataSourceName) {
...
-        SpanInScope spanInScope = tracer.withSpanInScope(resultSetSpan);
-        nestedResultSetsByStatement.computeIfAbsent(statementKey, key -> new HashSet<>()).add(resultSetKey);
-        resultSetSpans.put(resultSetKey, new SimpleEntry<>(resultSetSpan, spanInScope));
+        SpanWithScope spanWithScope = new SpanWithScope(resultSetSpan, tracer.withSpanInScope(resultSetSpan));
+        connectionInfo.getNestedResultSetSpans().put(resultSetKey, spanWithScope);
+
+        StatementInfo statementInfo = connectionInfo.getNestedStatements().get(statementKey);
+        // StatementInfo may be null when Statement is proxied and instance returned from ResultSet is different from instance returned in query method
+        // in this case if Statement is closed before ResultSet span won't be finished immediately, but when Connection is closed
+        if (statementInfo != null) {
```
Comment (reta) [inline datasource-decorator-spring-boot-autoconfigure/src/main/java/com/github/gavlyukovskiy/cloud/sleuth/TracingListenerStrategy.java]: :+1:

File: datasource-decorator-spring-boot-autoconfigure/src/main/java/com/github/gavlyukovskiy/cloud/sleuth/TracingListenerStrategy.java [hunk 2]
```diff
@@ -30,136 +21,203 @@ void beforeGetConnection(CON connectionKey, String dataSourceName) {
...
     }
 
     void afterGetConnection(CON connectionKey, Throwable t) {
-        Span connectionSpan = connectionSpans.get(connectionKey);
         if (t != null) {
-            connectionSpans.remove(connectionKey);
-            connectionSpan.kind(Kind.CLIENT);
-            connectionSpan.finish();
+            ConnectionInfo connectionInfo = openConnections.remove(connectionKey);
+            connectionInfo.getSpan().finish();
```
Comment (reta) [inline datasource-decorator-spring-boot-autoconfigure/src/main/java/com/github/gavlyukovskiy/cloud/sleuth/TracingListenerStrategy.java]: This one is pretty minor, do you think it should propagate error here?
```
        if (t != null) {
            connectionSpan.getSpan().error(t);
        }
```

**Comment by reta** [comment]
@gavlyukovskiy thanks a lot for the PR, it look good to me, the leak is gone, the spans are properly finished, I don't see dangling ones. The trace structure is looking good as well.

File: datasource-decorator-spring-boot-autoconfigure/src/main/java/com/github/gavlyukovskiy/cloud/sleuth/TracingListenerStrategy.java (hunk 2 above)
Comment (gavlyukovskiy) [reply datasource-decorator-spring-boot-autoconfigure/src/main/java/com/github/gavlyukovskiy/cloud/sleuth/TracingListenerStrategy.java]: Good catch! Added

**Pull request merged**